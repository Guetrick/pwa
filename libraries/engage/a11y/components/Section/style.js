import { css } from 'glamor';

export const hidden = css({
  display: 'none',
}).toString();
