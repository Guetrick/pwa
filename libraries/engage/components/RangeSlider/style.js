import { css } from 'glamor';

const outerRange = css({
  minHeight: 1,
  position: 'relative',
}).toString();

const range = css({
  left: 0,
  right: 0,
  position: 'absolute',
}).toString();

export default {
  outerRange,
  range,
};
