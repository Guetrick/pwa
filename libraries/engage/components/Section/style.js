import { css } from 'glamor';

export const hidden = css({
  display: 'none',
});

export const headline = css({
  clip: 'rect(1px, 1px, 1px, 1px)',
  height: '1px',
  margin: 0,
  overflow: 'hidden',
  padding: 0,
  position: 'absolute',
  whiteSpace: 'nowrap',
  width: '1px',
  zIndex: -1000,
});
