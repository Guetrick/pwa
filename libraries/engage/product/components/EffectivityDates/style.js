import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

export const hint = css({
  color: themeConfig.colors.success,
  fontSize: '0.875rem',
}).toString();
