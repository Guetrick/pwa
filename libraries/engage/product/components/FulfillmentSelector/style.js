import { css } from 'glamor';
import classNames from 'classnames';
import radioGroupStyles from '@shopgate/pwa-ui-shared/Form/RadioGroup/components/Item/style';

export const container = css({
  padding: '13px 16px 16px',
}).toString();

export const title = css({
  fontSize: 16,
  fontWeight: 500,
  marginBottom: 8,
}).toString();

export const radioGroup = css({
  // Removes the vertical padding, which are applied by default around radio groups
  paddingTop: 0,
  paddingBottom: 0,
}).toString();

export const radioItem = css({
  overflow: 'hidden',
  whiteSpace: 'pre',
}).toString();

export const radioItemLabelSpacer = css({
  whiteSpace: 'pre',
}).toString();

export const radioGroupItemLabel = classNames(radioGroupStyles.label, css({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
  alignItems: 'baseline',
}).toString());
