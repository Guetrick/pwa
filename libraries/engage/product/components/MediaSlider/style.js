import { css } from 'glamor';

export const full = css({
  width: '100%',
  height: '100%',
});

export const container = css({
  ...full,
  position: 'relative',
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
}).toString();
