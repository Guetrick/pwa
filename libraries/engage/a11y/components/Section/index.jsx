import React, {
  useRef, useState, useMemo, useLayoutEffect,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import kebabCase from 'lodash/kebabCase';
import { I18n } from '@shopgate/engage/components';
import { hidden } from './style';

/**
 * Checks the section ref has suitable child nodes.
 * @param {Object} ref A React ref.
 * @param {string} headlineId The ID of the section headline.
 * @returns {bool}
*/
const hasChildNodes = (ref, headlineId) => {
  if (!ref.current) {
    return false;
  }

  // Determine all child nodes except the headline.
  const childNodes = [].filter.call(
    ref.current.childNodes,
    el => el.getAttribute('id') !== headlineId
  );

  return childNodes.length > 0;
};

/**
 * The Section component is supposed to be used to structure the content to improve a11y. It
 * renders a headline on top which is only visible for screen readers and describes the section.
 * Internally a MutationObserver maintains the visibility based on the presence of rendered content.
 * @param {string} title The section title - can be a translation placeholder.
 * @param {Object} [titleParams={}] Additional parameters for the title  translation placeholder.
 * @param {Object} [className=null] A class name for the section.
 * @param {NodeList} [children=null] Component children.
 * @returns {JSX}
 */
const Section = ({
  title, titleParams, children, className, ...rest
}) => {
  const contentRef = useRef(null);
  const [hasContent, setHasContent] = useState(false);
  const id = useMemo(() => kebabCase(title), [title]);

  const observer = useMemo(() => new MutationObserver(() => {
    setHasContent(hasChildNodes(contentRef, id));
  }));

  useLayoutEffect(() => {
    setHasContent(hasChildNodes(contentRef, id));
    observer.observe(contentRef.current, { childList: true });

    return () => {
      observer.disconnect();
    };
  }, [contentRef, id, observer]);

  const classes = classNames(className, {
    [hidden]: !hasContent,
  });

  return (
    <section {...rest} ref={contentRef} className={classes} aria-labelledby={id}>
      <h1 id={id} hidden>
        <I18n.Text string={title} params={titleParams} />
      </h1>
      {children}
    </section>
  );
};

Section.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  titleParams: PropTypes.shape(),
};

Section.defaultProps = {
  children: null,
  className: null,
  titleParams: {},
};

export default Section;
