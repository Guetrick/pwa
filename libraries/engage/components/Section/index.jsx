import React, { useRef, useState, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { I18n } from '../';
import { headline, hidden } from './style';

/**
 * Checks the section ref has suitable child nodes.
 * @param {Object} ref A React ref.
 * @returns {bool}
 */
const hasChildNodes = (ref) => {
  if (!ref.current) {
    return false;
  }

  // Determine all child nodes except the headline.
  const childNodes = [].filter.call(
    ref.current.childNodes,
    el => !el.classList.contains(headline.toString())
  );

  return childNodes.length > 0;
};

/**
 * The Section component is supposed to be used to structure the content to improve a11y. It
 * renders a headline on top which is only visible for screen readers and describes the section.
 * Internally a MutationObserver maintains the visibility based on the presence of rendered content.
 * @param {string} title The section title - can be a translation placeholder.
 * @param {Object} titleParams Additional parameters for the title  translation placeholder.
 * @param {NodeList} children Component children.
 * @returns {JSX}
 */
const Section = ({ title, titleParams, children }) => {
  const contentRef = useRef(null);
  const [hasContent, setHasContent] = useState(false);

  const observer = useMemo(() => new MutationObserver(() => {
    setHasContent(hasChildNodes(contentRef));
  }));

  useEffect(() => {
    setHasContent(hasChildNodes(contentRef));
    observer.observe(contentRef.current, { childList: true });

    return () => {
      observer.disconnect();
    };
  }, [contentRef]);

  return (
    <section ref={contentRef} className={hasContent ? null : hidden}>
      <h1 className={headline}>
        <I18n.Text string={title} params={titleParams} />
      </h1>
      {children}
    </section>
  );
};

Section.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
  titleParams: PropTypes.shape(),
};

Section.defaultProps = {
  children: null,
  titleParams: {},
};

export default Section;
