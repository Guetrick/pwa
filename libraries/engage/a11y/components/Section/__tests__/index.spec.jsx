/* eslint-disable require-jsdoc, extra-rules/potential-point-free,
class-methods-use-this, extra-rules/no-single-line-objects */
import React from 'react';
import { mount } from 'enzyme';
import Section from '../index';
import { hidden } from '../style';

const mutationConstructorSpy = jest.fn();
const mutationObserveSpy = jest.fn();
const mutationDisconnectSpy = jest.fn();

global.MutationObserver = class {
  constructor(callback) { mutationConstructorSpy(callback); }
  observe(element, initObject) { mutationObserveSpy(element, initObject); }
  disconnect() { mutationDisconnectSpy(); }
};

const mockTranslate = jest.fn();

jest.mock('../../../../components', () => ({
  I18n: {
    Text: function Translate(props) {
      mockTranslate(props);
      return props.string;
    },
  },
}));

const title = 'my.headline';
const titleParams = { some: 'params' };

describe('<Section />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render visible', () => {
    const wrapper = mount((
      <Section title={title} titleParams={titleParams}>
        <div id="child-component" />
      </Section>
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('section').hasClass(hidden.toString())).toBe(false);
    expect(mockTranslate).toHaveBeenCalledWith({ string: title, params: titleParams });
  });

  it('should render hidden when no children rendered', () => {
    const wrapper = mount(<Section title={title} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('section').hasClass(hidden.toString())).toBe(true);
    expect(mockTranslate).toHaveBeenCalledWith({ string: title, params: {} });
  });

  describe('MutationObserver', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(<Section title={title} />);
    });

    it('should create an observer instance', () => {
      expect(mutationConstructorSpy).toHaveBeenCalledTimes(1);
      expect(mutationConstructorSpy).toHaveBeenCalledWith(expect.any(Function));
    });

    it('should observe with the correct initialization', () => {
      expect(mutationObserveSpy).toHaveBeenCalledTimes(1);
      expect(mutationObserveSpy).toHaveBeenCalledWith(
        wrapper.find('section').instance(),
        {
          childList: true,
        }
      );
    });

    it('should call disconnect on unmount', () => {
      wrapper.unmount();
      expect(mutationDisconnectSpy).toHaveBeenCalledTimes(1);
    });
  });
});

/* eslint-enable require-jsdoc, extra-rules/potential-point-free,
class-methods-use-this, extra-rules/no-single-line-objects */
