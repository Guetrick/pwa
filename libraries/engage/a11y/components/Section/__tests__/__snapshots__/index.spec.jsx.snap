// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section /> should render hidden when no children rendered 1`] = `
<Section
  className={null}
  title="my.headline"
  titleParams={Object {}}
>
  <section
    aria-labelledby="my-headline"
    className="css-gyrv08"
  >
    <h1
      hidden={true}
      id="my-headline"
    >
      <Translate
        params={Object {}}
        string="my.headline"
      >
        my.headline
      </Translate>
    </h1>
  </section>
</Section>
`;

exports[`<Section /> should render visible 1`] = `
<Section
  className={null}
  title="my.headline"
  titleParams={
    Object {
      "some": "params",
    }
  }
>
  <section
    aria-labelledby="my-headline"
    className="css-gyrv08"
  >
    <h1
      hidden={true}
      id="my-headline"
    >
      <Translate
        params={
          Object {
            "some": "params",
          }
        }
        string="my.headline"
      >
        my.headline
      </Translate>
    </h1>
    <div
      id="child-component"
    />
  </section>
</Section>
`;
